### DISCUSSION TITLE
Optional Tuple with Custom Structs throws Segmentation Error

### DISCUSSION THREAD

**Discussion by ghost** [modular/modular#1752] [Q&A]
I have a struct `Scalar` that should take in an integer value, and then _optionally_ two more `Scalar`s. Here's my current implementation:

```py
from collections.optional import Optional


struct Scalar:
    var _children: Optional[Tuple[Scalar, Scalar]]

    fn __init__(
        inout self,
        _children: Optional[Tuple[Scalar, Scalar]],
    ) -> None:
        self._children = _children
```

If the contents of the `Tuple`s were `Int`s, this works fine - but as soon as I change them to `Scalar`s, Mojo crashes with a segmentation error, even though all types are valid.

How can I fix this, or is there some other way to implement this? Thanks in advance.

**Edit:**
See [this reply](https://github.com/modularml/mojo/discussions/1752?sort=top#discussioncomment-8494667) for details.

**Comment by Brian-M-J** [comment]
Have a look at [`Optional`](https://docs.modular.com/mojo/stdlib/collections/optional.html#optional). I don't think an empty tuple (`()`) would work with it but `(None, None)` would.

**Reply by ghost** [reply]
Thanks, that definitely helped. I was able to use something like this, which fixes my earlier issue:

```py
from collections.optional import Optional


struct Scalar:
    var _children: Optional[Tuple[Scalar, Scalar]]

    fn __init__(
        inout self,
        _children: Optional[Tuple[Scalar, Scalar]],
    ) -> None:
        self._children = _children
```

While using `Optional[Tuple[Scalar, Scalar]]` works fine for the `_children` variable, when I change the `__init__()` parameter, however, Mojo crashes with a segmentation fault. How can I fix that? Thanks so much for helping.

**Reply by Brian-M-J** [reply]
I'm not a 100% sure what's going on, but in the docs for `Optional` it says:
>Currently T is required to be a `CollectionElement` so we can implement copy/move for Optional and allow it to be used in collections itself.

Your `Scalar` struct does not implement the [`CollectionElement`](https://docs.modular.com/mojo/stdlib/collections/vector.html#collectionelement) trait, which may be what's causing the issue.

Try adding the [`@value`](https://docs.modular.com/mojo/manual/decorators/value.html) decorator to the struct and make it inherit from `CollectionElement` as demonstrated [here](https://docs.modular.com/mojo/manual/traits#using-traits).

Edit: Also while looking through the [`Tuple`](https://docs.modular.com/mojo/stdlib/builtin/tuple.html#tuple) docs:
>The type of a literal tuple expression.

Since the `Tuple` type is meant for literal expressions, it may not be able to be stored in structs or variables. Try using a [`StaticTuple`](https://docs.modular.com/mojo/stdlib/utils/static_tuple.html#statictuple) instead.

**Reply by ghost** [reply]
I tried playing around with it a bit, what I found was that the following worked:
```py
from collections.optional import Optional


struct Scalar:
    var _children: Optional[Tuple[Int, Int]]

    fn __init__(
        inout self,
        _children: Optional[Tuple[Int, Int]],
    ) -> None:
        self._children = _children
```

But as soon as I changed from `Int`s to `Scalar`s, Mojo threw a segmentation fault without any clear error message. As for the `Optional` type requiring a `CollectionElement`, that should have still worked because the `Scalar`s are inside a `Tuple` - `Tuple` itself should be able to store any type.

I tried out the `StaticTuple` instead, but unfortunately since that doesn't have the `CollectionElement` trait I can't use it with `Optional`, so I'm kind of lost on what else I can do.

If you have any other ideas, they would be greatly appreciated - I've also been thinking of some other way I can implement this, but I don't know anything else (`Scalar` needs to take in an integer, and then optionally two other `Scalar`s).

Thank you so much for all the help.
